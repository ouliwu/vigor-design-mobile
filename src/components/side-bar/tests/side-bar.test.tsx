import React, { useState } from 'react'
import { render, testA11y, fireEvent } from 'testing'
import SideBar from '..'
import { Badge } from 'vigor-design-mobile'

const classPrefix = `adm-side-bar`

const tabs = [
  {
    key: 'key1',
    title: 'item1',
    badge: Badge.dot,
  },
  {
    key: 'key2',
    title: 'item2',
    badge: '5',
  },
  {
    key: 'key3',
    title: 'item3',
    badge: '99+',
    disabled: true,
  },
]
describe('TabBar', () => {
  test('a11y', async () => {
    await testA11y(
      <SideBar>
        {tabs.map(item => (
          <SideBar.Item key={item.key} title={item.title} />
        ))}
      </SideBar>
    )
  })

  test('basic usage', async () => {
    const onChange = jest.fn()
    const { getByText } = render(
      <SideBar onChange={onChange} defaultActiveKey='key2'>
        {tabs.map(item => (
          <SideBar.Item key={item.key} title={item.title} badge={item.badge} />
        ))}
      </SideBar>
    )

    const items = document.querySelectorAll(`.${classPrefix}-item`)

    expect(items[1]).toHaveClass(`${classPrefix}-item-active`)
    fireEvent.click(getByText('item3'))
    expect(items[2]).toHaveClass(`${classPrefix}-item-active`)
    expect(onChange).toBeCalledWith('key3')
  })

  test('controlled mode', async () => {
    const App = () => {
      const [activeKey, setActiveKey] = useState('key3')
      return (
        <SideBar activeKey={activeKey} onChange={setActiveKey}>
          {tabs.map(item => (
            <SideBar.Item key={item.key} title={item.title} />
          ))}
        </SideBar>
      )
    }
    const { getByText } = render(<App />)
    const items = document.querySelectorAll(`.${classPrefix}-item`)
    expect(items[2]).toHaveClass(`${classPrefix}-item-active`)
    fireEvent.click(getByText('item1'))
    expect(items[0]).toHaveClass(`${classPrefix}-item-active`)
  })

  test('disabled item', async () => {
    const onChange = jest.fn()
    const { getByText } = render(
      <SideBar onChange={onChange}>
        {tabs.map(item => (
          <SideBar.Item
            key={item.key}
            title={item.title}
            disabled={item.disabled}
          />
        ))}
      </SideBar>
    )

    const items = document.querySelectorAll(`.${classPrefix}-item`)
    expect(items[2]).toHaveClass(`${classPrefix}-item-disabled`)
    fireEvent.click(getByText('item3'))
    expect(onChange).not.toBeCalled()
  })

  test('the children should be react valid element', async () => {
    render(<SideBar>{1}</SideBar>)
    expect(
      document.querySelectorAll(`.${classPrefix}-items`)[0]
    ).toBeEmptyDOMElement()
  })

  test('item content', async () => {
    const { container } = render(
      <SideBar>
        <SideBar.Item key='title' title='title' />
        <SideBar.Item key='empty' />
        <SideBar.Item />
      </SideBar>
    )

    expect(container).toMatchSnapshot()
  })
})
