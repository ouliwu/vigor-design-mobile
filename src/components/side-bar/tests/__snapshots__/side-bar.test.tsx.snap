// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar item content 1`] = `
<div>
  <div
    class="adm-side-bar"
  >
    <div
      class="adm-side-bar-items"
    >
      <div
        class="adm-side-bar-item adm-side-bar-item-active"
      >
        <div
          class="adm-badge-wrapper"
        >
          <div
            class="adm-side-bar-item-title"
          >
            <div
              class="adm-side-bar-item-highlight"
            />
            title
          </div>
        </div>
      </div>
      <div
        class="adm-side-bar-item"
      >
        <svg
          class="adm-side-bar-item-corner adm-side-bar-item-corner-top"
          viewBox="0 0 30 30"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            stroke="none"
            stroke-width="1"
          >
            <path
              d="M30,0 C13.4314575,3.04359188e-15 -2.02906125e-15,13.4314575 0,30 L0,30 L0,0 Z"
              fill="var(--adm-color-background)"
              transform="translate(15.000000, 15.000000) scale(-1, -1) translate(-15.000000, -15.000000) "
            />
          </g>
        </svg>
        <div
          class="adm-badge-wrapper"
        >
          <div
            class="adm-side-bar-item-title"
          />
        </div>
      </div>
    </div>
    <div
      class="adm-side-bar-extra-space"
    />
  </div>
</div>
`;
